Various sanitycheck fixes and improvements..

nashif opened the pull request:
use python multiprocessing instead of a thread pool, resolves
various issues.


- [x] Device testing support with multiprocessing

Fixes #28040
Fixes #26949
Fixes #24652
Fixes #30099
Fixes #27583

nashif committed:
sanitycheck: print pid of qemu process when debugging

For debugging, print the PID of the qemu running.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: use multiprcoessing instead of threads

With python threading we have multiple issues with performance and
leakage. Use the python Process and implement locking using
multiprocessing library. This change improves performance and fixes
various issues with logging, concurrency and reliability of the output.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

andrewboie commented:
can we rebase this?

galak commented:
@nashif do we want to pull some of the changes out and look to merge while you work on the device testing?

nashif commented:
> @nashif do we want to pull some of the changes out and look to merge while you work on the device testing?

it is all or nothing unfortunately.

andrewboie commented:
naturally all the FIXMEs and stuff need to be addressed but I'm getting good results testing this.
One thing I did notice is that there is still a long delay at the end, after (for example) `INFO    - 804 tests executed on platforms, 162 tests were only built.` Seems unreasomable for a modern CPU to be spinning for a long time when only 1000 tests were run.

PerMac commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -84,7 +86,176 @@
+
+    def get_skipped_cases(self):
+        with self.lock:
+            return self.skipped_cases.value
+
+    def get_done(self):
+        with self.lock:
+            return self.done.value
+
+    def get_passed(self):
+        with self.lock:
+            return self.passed.value
+
+    def get_skipped(self):
+        with self.lock:
+            return self.skipped.value
+
+    def get_runtime_skipped(self):
+        with self.lock:
+            return self.runtime_skipped.value
+
+    def get_failed(self):
+        with self.lock:
+            return self.failed.value
+
+    def get_error(self):
+        with self.lock:
+            return self.error.value
+
+    def get_cases(self):
+        with self.lock:
+            return self.cases.value
+
+    def get_total(self):
+        with self.lock:
+            return self.total.value
+
+    def set_skipped(self, value):
+        with self.lock:
+            self.skipped.value = value
```
Repeated.

PerMac commented on scripts/sanitycheck:
```diff
@@ -1138,17 +1151,28 @@ def main():
-            suite.total_done = suite.total_tests - suite.total_failed
-            suite.total_failed = suite.total_errors
+            # FIXME
+            done1 = results.get_total() - results.get_failed()
+            failed1 = results.get_error()
```
I see FIXME comment, but just in case: `done1` and `failed1` are used nowhere. What is the plan for them?

PerMac commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -2119,18 +2317,22 @@ def process(self, message):
             pipeline.put({
                 "op": "report",
                 "test": self.instance,
-                "state": "executed",
```
I can't find why this is removed and if there will be any consequences. Can you explain this change?

nashif replied:
it is no being used anywhere.

PerMac commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -3009,88 +3189,52 @@ def add_tasks_to_queue(self, build_only=False, test_only=False):
         logger.info("Adding tasks to the queue...")
-        # We can use a with statement to ensure threads are cleaned up promptly
-        with BoundedExecutor(bound=self.jobs, max_workers=self.jobs) as executor:
-
```
The removed code below has comments explaining what was going on, and it seems the new code lacks commenting.  I would really appreciate it if you would add some comments in the code explaining what is going on there.

PerMac reviewed (dismissed):
I am not able to verify the changes related to the processing/threading and the change for coverage, this is beyond my experience. I will run some checks to see how the new counting is performing.

nashif committed:
sanitycheck: add some debugging during report phase

Show in debug mode what files are being written. In some cases this
takes a while and it is good to see what is going on.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: reduce verbosity

Reduce debug messages while scanning directories for tests.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: fix arch filtering

If we provide arch on command line, do not check for all platforms of
architectures not specified.

Fixes #30099

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -2464,24 +2633,19 @@ def info(what):
             if instance.status == 'skipped':
-                self.total_skipped += 1
-                self.total_skipped_cases += len(instance.testcase.cases)
+                results.increment_skipped()
```
with multiprocessing it gets more complex, will look into that and optimize one we have the basics working.

nashif replied:
done now.

nashif committed:
sanitycheck: error out if we do not find any tests

If we find no tests, error out instead of exiting silently.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: do not report about execution if --build-only

When running with --build-only, do not print messages about tests being
executed.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: report test name if no status

Also catch cases where we try to test on a device when no device is
connected.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@PerMac  if you can test this with --device-testing, would be appreciated.

nashif committed:
sanitycheck: unify terminology and fix accounting

Try and keep terminilogy consistent, a test application with multiple
test cases is now a 'test suite'.

Also fix accounting when retrying failed.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

PerMac commented:
I think something is wrong with --device-testing when hardware map is used. I used a setup with 2 platforms connected (out of several from 4 types of boards). When I run hello_word it tells that 4 configs will be tested, 2 pass and then it hangs waiting for output from not connected types of boards. It seems timeout is also not working for such configuration as I had to kill it after several minutes.

nashif commented:
> I think something is wrong with --device-testing when hardware map is used. I used a setup with 2 platforms connected (out of several from 4 types of boards). When I run hello_word it tells that 4 configs will be tested, 2 pass and then it hangs waiting for output from not connected types of boards. It seems timeout is also not working for such configuration as I had to kill it after several minutes.

how did you invoke sanitycheck exactly?

nashif committed:
sanitycheck: schema: 'available' is a runtime key

The available key is a runtime variable, it does not need to be in the
hardware map. Make it optional to keep the file format compatible for
now.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: cleanup hardware map handling

Shorten class method names and print hardware map as before after
scanning.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

PerMac commented:
`./scripts/sanitycheck --device-testing --hardware-map map.yml -T samples/hello_world/`
when called in master:

<details><summary>expand</summary>
<p>

./scripts/sanitycheck --device-testing --hardware-map map.yml -T samples/hello_world/
ZEPHYR_BASE unset, using "/home/maciej/zephyrproject2/zephyr"
Renaming output directory to /home/maciej/zephyrproject2/zephyr/sanity-out.79
INFO    - Zephyr version: zephyr-v2.4.0-1798-g1d4c36486c
INFO    - JOBS: 4
INFO    - Building initial testcase list...
INFO    - 1 test configurations selected, 0 configurations discarded due to filters.

Device testing on:

| Platform            |           ID | Serial device   |
|---------------------|--------------|-----------------|
| nrf52840dk_nrf52840 | 000683788957 |                 |
| nrf52840dk_nrf52840 | 000683810635 |                 |
| nrf52840dk_nrf52840 | 000683217973 | /dev/ttyACM0    |

INFO    - Adding tasks to the queue...
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0
INFO    - 1 of 1 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 9.15 seconds
INFO    - In total 1 test cases were executed on 1 out of total 309 platforms (0.32%)
INFO    - 1 tests executed on platforms, 0 tests were only built.

Hardware distribution summary:

| Board               |           ID |   Counter |
|---------------------|--------------|-----------|
| nrf52840dk_nrf52840 | 000683217973 |         1 |

</p>
</details>


When called in  within your PR with the same hardware map (had to kill it):

<details><summary>expand</summary>
<p>
$ ./scripts/sanitycheck --device-testing --hardware-map map.yml -T samples/hello_world/
ZEPHYR_BASE unset, using "/home/maciej/zephyrproject2/zephyr"
Renaming output directory to /home/maciej/zephyrproject2/zephyr/sanity-out.80
INFO    - Zephyr version: zephyr-v2.4.0-1833-gfb44f1d87a
INFO    - JOBS: 4
INFO    - Building initial testcase list...
INFO    - 1 test suites (5 configurations) selected, 0 configurations discarded due to filters.

Device testing on:

| Platform                  |           ID | Serial device   |
|---------------------------|--------------|-----------------|
| nrf52840dk_nrf52840       | 000683788957 |                 |
| nrf5340dk_nrf5340_cpuapp  | 000960146054 |                 |
| nrf9160dk_nrf9160         | 000960033217 |                 |
| nrf52840dk_nrf52840       | 000683810635 |                 |
| nrf52dk_nrf52832          | 000682923426 |                 |
| nrf52840dk_nrf52840       | 000683217973 | /dev/ttyACM0    |
| nrf5340pdk_nrf5340_cpuapp | 000960177695 |                 |

INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:    1/   5  20%  skipped:    0, failed:    0^CTraceback (most recent call last):
  File "./scripts/sanitycheck", line 1245, in <module>
    main()
  File "./scripts/sanitycheck", line 1158, in main
    results = suite.execute(pipeline, done_queue, results)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 3214, in execute
Process Process-2:
    p.join()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 124, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python3.6/multiprocessing/popen_fork.py", line 50, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/usr/lib/python3.6/multiprocessing/popen_fork.py", line 28, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt
Traceback (most recent call last):
Process Process-4:
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 3196, in pipeline_mgr
    pb.process(pipeline, done_queue, task, lock, results)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2269, in process
    self.run()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2479, in run
    instance.handler.handle()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 688, in handle
    time.sleep(1)
KeyboardInterrupt
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 3196, in pipeline_mgr
    pb.process(pipeline, done_queue, task, lock, results)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2269, in process
    self.run()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2479, in run
    instance.handler.handle()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 688, in handle
    time.sleep(1)
KeyboardInterrupt
Process Process-5:
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 3196, in pipeline_mgr
    pb.process(pipeline, done_queue, task, lock, results)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2269, in process
    self.run()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2479, in run
    instance.handler.handle()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 688, in handle
    time.sleep(1)
KeyboardInterrupt
Process Process-3:
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 3196, in pipeline_mgr
    pb.process(pipeline, done_queue, task, lock, results)
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2269, in process
    self.run()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 2479, in run
    instance.handler.handle()
  File "/home/maciej/zephyrproject2/zephyr/scripts/sanity_chk/sanitylib.py", line 688, in handle
    time.sleep(1)
KeyboardInterrupt
</p>
</details>


hardware map: 
<details><summary>map.yml</summary>
- available: true
  connected: false
  id: 000683788957
  platform: nrf52840dk_nrf52840
  product: J-Link
  runner: nrfjprog
  serial: null
- available: true
  connected: false
  id: 000683188034
  platform: unknown
  product: J-Link
  runner: jlink
  serial: null
- available: true
  connected: false
  id: 000960146054
  platform: unknown
  product: J-Link
  runner: jlink
  serial: null
- available: true
  connected: false
  id: 000960146054
  platform: unknown
  product: J-Link
  runner: jlink
  serial: null
- available: true
  connected: false
  id: 000960146054
  platform: nrf5340dk_nrf5340_cpuapp
  product: J-Link
  runner: nrfjprog
  serial: null
- available: true
  connected: false
  id: 000960033217
  platform: nrf9160dk_nrf9160
  product: J-Link
  runner: nrfjprog
  serial: null
- available: true
  connected: false
  id: 000960033217
  platform: unknown
  product: J-Link
  runner: jlink
  serial: null
- available: true
  connected: false
  id: 000960033217
  platform: unknown
  product: J-Link
  runner: jlink
  serial: null
- available: true
  connected: false
  id: 000683810635
  platform: nrf52840dk_nrf52840
  product: J-Link
  runner: nrfjprog
  serial: null
- available: true
  connected: false
  id: 000682923426
  platform: nrf52dk_nrf52832
  product: J-Link
  runner: nrfjprog
  serial: null
- available: true
  connected: true
  id: 000683217973
  platform: nrf52840dk_nrf52840
  product: J-Link
  runner: nrfjprog
  serial: /dev/ttyACM0
- available: true
  connected: false
  id: 000960177695
  platform: unknown
  product: J-Link
  runner: jlink
  serial: null
- available: true
  connected: false
  id: 000960177695
  platform: unknown
  product: J-Link
  runner: jlink
  serial: null
- available: true
  connected: false
  id: 000960177695
  platform: nrf5340pdk_nrf5340_cpuapp
  product: J-Link
  runner: jlink
  serial: null

</details>

PerMac commented:
It is interesting also how the table "Device testing on" in the output looks like. In both cases, it has excess information. In the master it shows multiple variants (also not connected ones) of the platform that is actually connected
In this PR it shows not connected ones of other types as well

nashif committed:
sanitycheck: tests: simplify filter testcase

use a loop instead of all() to make the test easier to debug.
remove the default platform test, as this is now being filtered
differently in the code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: fix hardware map defaults

We were setting all devices as connected, now do that based on serial
availability.
Rename internal variables and make them shorter:

connected_hardware -> duts
ConnectedDevice = DUT

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@PerMac thanks, found the issue and fixed it hopefully.

nashif committed:
sanitycheck: add option to treat overflows as errors

Add a new option that will enable treating RAM/SRAM overflows as errors
instead of skipping them, which is the default.

Fixes #27583

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
> One thing I did notice is that there is still a long delay at the end, after (for example) `INFO - 804 tests executed on platforms, 162 tests were only built.` Seems unreasomable for a modern CPU to be spinning for a long time when only 1000 tests were run.

I suspect this is the time it takes to write reports and probably related to #30099 which is now fixed. We were basically writing reports for completely irrelevant platforms.

maksimmasalski commented:
@nashif Sorry, I saw that late, it was a midnight in my town when you requested a code review.

evgeniy-paltsev commented:
@nashif 
The 531fe89e80f859d9bc820418059cec93463e0539 commit of this series causes various regressions for ARC.

One of the regressions is that we get some exception from sanitycheck code and mark test as FAILED however it successfully passed.

Here is an example on exception:
```
Exception in thread Thread-15603:
Traceback (most recent call last):
  File "/home/paltsev/CodeData/zephyr/zephyr-arc-64/zephyr/scripts/sanity_chk/sanitylib.py", line 496, in _output_handler
    proc.wait(0.1)
  File "/usr/lib64/python3.7/subprocess.py", line 1019, in wait
    return self._wait(timeout=timeout)
  File "/usr/lib64/python3.7/subprocess.py", line 1645, in _wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['make', 'run']' timed out after 0.1 seconds

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/paltsev/CodeData/zephyr/zephyr-arc-64/zephyr/scripts/sanity_chk/sanitylib.py", line 498, in _output_handler
    self.terminate(proc)
  File "/home/paltsev/CodeData/zephyr/zephyr-arc-64/zephyr/scripts/sanity_chk/sanitylib.py", line 456, in terminate
    os.kill(child.pid, signal.SIGTERM)
ProcessLookupError: [Errno 3] No such process

ERROR   - nsim_em                   tests/kernel/mutex/sys_mutex/system.mutex          FAILED: Failed
ERROR   - see: /home/paltsev/CodeData/zephyr/zephyr-arc-64/zephyr/sanity-out/nsim_em/tests/kernel/mutex/sys_mutex/system.mutex/handler.log
```

In the `tests/kernel/mutex/sys_mutex/system.mutex/handler.log` everything looks OK:
```
 [-snip-]
[100%] Built target zephyr_final
Scanning dependencies of target run
Console now belongs to UART, hit CRTL-] to return to simulator.
*** Booting Zephyr OS build zephyr-v2.4.0-1942-g531fe89e80f8  ***
Running test suite mutex_complex
===================================================================
START - test_mutex
START - Test kernel Mutex API
===================================================================
Done LOCKING!  Current priority = 5
Testing recursive locking
Recursive locking tests successful
 PASS - test_mutex
===================================================================
START - test_user_access
E: syscall check_sys_mutex_addr failed check: Memory region 0x80000c08 (size 4) write access denied
E: syscall check_sys_mutex_addr failed check: Memory region 0x80000c08 (size 4) write access denied
 PASS - test_user_access
===================================================================
START - test_supervisor_access
 PASS - test_supervisor_access
===================================================================
Test suite mutex_complex succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Zephyr SDK is 0.11.4, toolchain is zephyr one.

How-to reproduce:
```
./scripts/sanitycheck --platform nsim_em
```

The tests fail randomly. Usually 2-5 tests fail for nsim_em.

Given that you have nSIM may I ask you to take a look at this?
Thanks.

abrodkin commented:
@evgeniy-paltsev could you please file an issue with details above and mention this PR there for tracking this problem especially in light of the release getting closer & closer?

evgeniy-paltsev commented:
Here is the issue for this: #30752
